File: backend/src/schemas/dataset.py
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class DatasetCreate(BaseModel):
    """
    Request schema for registering dataset metadata.
    """

    model_config = ConfigDict(populate_by_name=True)

    name: str = Field(min_length=1, max_length=255)
    original_filename: str = Field(min_length=1, max_length=255)
    storage_format: str = "csv"
    content_type: str = "text/csv"
    file_size_bytes: int = Field(ge=0)
    row_count: int = Field(ge=0)
    column_count: int = Field(ge=0)
    dataset_schema: dict[str, Any] = Field(alias="schema_json")


class DatasetRead(BaseModel):
    """
    Response schema for a dataset.
    """

    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True,
    )

    id: UUID
    name: str
    original_filename: str
    s3_key: str
    storage_format: str
    content_type: str
    file_size_bytes: int
    row_count: int
    column_count: int
    dataset_schema: dict[str, Any] = Field(alias="schema_json")
    status: str
    query_count: int
    last_query_at: datetime | None
    created_at: datetime
    updated_at: datetime


class DatasetPreviewResponse(BaseModel):
    """
    Response schema for previewing dataset rows.
    """

    dataset_id: UUID
    columns: list[str]
    rows: list[dict[str, Any]]
    row_count: int
    limit: int
    execution_time_ms: float


class DatasetDownloadUrlResponse(BaseModel):
    """
    Response schema for temporary dataset download URLs.
    """

    dataset_id: UUID
    object_key: str
    download_url: str
    expires_in_seconds: int
    storage_backend: str


class DatasetListResponse(BaseModel):
    """
    Response schema for listing datasets.
    """

    items: list[DatasetRead]
    count: int
    limit: int
    offset: int

File: frontend/src/types/dataset.ts
export type DatasetSchema = Record<string, string>;

export type Dataset = {
  id: string;
  name: string;
  original_filename: string;
  s3_key: string;
  storage_format: string;
  content_type: string;
  file_size_bytes: number;
  row_count: number;
  column_count: number;
  schema_json?: DatasetSchema;
  dataset_schema?: DatasetSchema;
  status: string;
  query_count: number;
  last_query_at: string | null;
  created_at: string;
  updated_at: string;
};

export type DatasetListResponse = {
  items: Dataset[];
  count: number;
  limit: number;
  offset: number;
};

export type DatasetPreviewResponse = {
  dataset_id: string;
  columns: string[];
  rows: Record<string, unknown>[];
  row_count: number;
  limit: number;
  execution_time_ms: number;
};

export type DatasetDownloadUrlResponse = {
  dataset_id: string;
  object_key: string;
  download_url: string;
  expires_in_seconds: number;
  storage_backend: string;
};
